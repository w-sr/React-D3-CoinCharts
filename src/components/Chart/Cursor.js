import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import { animation, color, size } from "../../styles/constants";

const StyledCursor = styled.svg.attrs({
  style: ({ visible }) => ({
    opacity: visible ? 1 : 0,
  }),
})`
  height: 100%;
  width: 100%;
  transition: opacity ${animation.speed};
`;

const Circle = styled.circle`
  r: ${size.small};
  fill: ${color.white};
  stroke: ${color.coinchartsGray};
  stroke-width: 2;
`;

const Line = styled.line`
  stroke: ${color.coinchartsGray};
  stroke-width: 1;
`;

const Cursor = ({ height, visible, x, y }) => (
  <StyledCursor data-testid="Cursor" visible={visible}>
    <Line x1={x} x2={x} y1={0} y2={height} />
    <Circle cx={x} cy={y} />
  </StyledCursor>
);

Cursor.propTypes = {
  height: PropTypes.number.isRequired,
  visible: PropTypes.bool.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
};

export default Cursor;
