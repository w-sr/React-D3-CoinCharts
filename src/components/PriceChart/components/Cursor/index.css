.Cursor.show {
  opacity: 1;
  transition: opacity 300ms;
}

.Cursor.hidden {
  opacity: 0;
  transition: opacity 300ms;
}

.Cursor .line {
  stroke: #7d95b6;
  stroke-width: 1;
}

.Cursor .circle {
  fill: white;
  stroke: #7d95b6;
  stroke-width: 2;
}
