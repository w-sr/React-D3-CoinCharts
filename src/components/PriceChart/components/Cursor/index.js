import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './index.css';

const CURSOR_RADIUS_SIZE = 4;

const Cursor = ({ height, visible, x, y }) => {
  const cursorClass = classNames({
    Cursor: true,
    hidden: !visible,
    show: visible,
  });

  return (
    <g className={cursorClass}>
      <line
        className="line"
        x1={x}
        x2={x}
        y1={0}
        y2={height}
      />
      <circle
        className="circle"
        cx={x}
        cy={y}
        r={CURSOR_RADIUS_SIZE}
      />
    </g>
  );
};

Cursor.propTypes = {
  height: PropTypes.number.isRequired,
  visible: PropTypes.bool.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
};

export default Cursor;
