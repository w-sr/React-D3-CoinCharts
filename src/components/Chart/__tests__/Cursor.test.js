import React from "react";
import { render } from "react-testing-library";

import Cursor from "../Cursor";

const TEST_PROPS = {
  height: 5,
  x: 10,
  y: 15,
  visible: true,
};

describe("<Cursor />", () => {
  it("renders without crashing", () => {
    const { container } = render(<Cursor {...TEST_PROPS} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders in correct position based on `props.x`, `props.y` & `props.height`", () => {
    const { container } = render(<Cursor {...TEST_PROPS} />);
    expect(container.querySelector("line").getAttribute("x1")).toEqual(`${TEST_PROPS.x}`);
    expect(container.querySelector("line").getAttribute("x2")).toEqual(`${TEST_PROPS.x}`);
    expect(container.querySelector("line").getAttribute("y1")).toEqual("0");
    expect(container.querySelector("line").getAttribute("y2")).toEqual(`${TEST_PROPS.height}`);
    expect(container.querySelector("circle").getAttribute("cx")).toEqual(`${TEST_PROPS.x}`);
    expect(container.querySelector("circle").getAttribute("cy")).toEqual(`${TEST_PROPS.y}`);
  });

  it("shows component when `props.visible` is true", () => {
    const { container } = render(<Cursor {...TEST_PROPS} visible />);
    expect(container.firstChild.style.opacity).toEqual("1");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("hides component when `props.visible` is false", () => {
    const { container } = render(<Cursor {...TEST_PROPS} visible={false} />);
    expect(container.firstChild.style.opacity).toEqual("0");
    expect(container.firstChild).toMatchSnapshot();
  });
});
