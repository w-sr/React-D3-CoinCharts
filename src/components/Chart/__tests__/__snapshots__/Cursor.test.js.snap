// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cursor /> hides component when \`props.visible\` is false 1`] = `
<svg
  style="opacity: 0;"
  class="sc-bdVaJa ltcOW"
  data-testid="Cursor"
>
  <line
    class="sc-htpNat jVvTZn"
    x1="10"
    x2="10"
    y1="0"
    y2="5"
  />
  <circle
    class="sc-bwzfXH byisMI"
    cx="10"
    cy="15"
  />
</svg>
`;

exports[`<Cursor /> renders without crashing 1`] = `
<svg
  style="opacity: 1;"
  class="sc-bdVaJa ltcOW"
  data-testid="Cursor"
>
  <line
    class="sc-htpNat jVvTZn"
    x1="10"
    x2="10"
    y1="0"
    y2="5"
  />
  <circle
    class="sc-bwzfXH byisMI"
    cx="10"
    cy="15"
  />
</svg>
`;

exports[`<Cursor /> shows component when \`props.visible\` is true 1`] = `
<svg
  style="opacity: 1;"
  class="sc-bdVaJa ltcOW"
  data-testid="Cursor"
>
  <line
    class="sc-htpNat jVvTZn"
    x1="10"
    x2="10"
    y1="0"
    y2="5"
  />
  <circle
    class="sc-bwzfXH byisMI"
    cx="10"
    cy="15"
  />
</svg>
`;
